fpga-arch/bsv/SWD.bsv

```diff
@@ -0,0 +1,188 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+//
+// Copyright 2024 Ledger SAS, written by Charles Christen
+
+import SWDInner::*;
+
+import GetPut::*;
+import Vector::*;
+import StmtFSM::*;
+import ClientServer::*;
+
+(* always_ready, always_enabled *)
+interface ScaffoldBus;
+    // bus_in_t
+    (* prefix="" *) method Action address((* port="address" *) Bit#(16) a);
+    (* prefix="" *) method Action write_data((* port="write_data" *) Bit#(8) w);
+    (* prefix="" *) method Action write((* port="write" *) Bit#(1) b);
+    (* prefix="" *) method Action read((* port="read" *) Bit#(1) b);
+
+    // register selection
+    (* prefix="" *) method Action en_rdata((* port="en_rdata" *) Bit#(1) en);
+    (* prefix="" *) method Action en_wdata((* port="en_wdata" *) Bit#(1) en);
+    (* prefix="" *) method Action en_cmd((* port="en_cmd" *) Bit#(1) en);
+    (* prefix="" *) method Action en_status((* port="en_status" *) Bit#(1) en);
+
+    // readable register out
+    (* prefix="" *) method Bit#(8) reg_rdata;
+    (* prefix="" *) method Bit#(8) reg_status;
+endinterface
+
+interface ScaffoldSWDModule;
+    (* prefix="" *) interface ScaffoldBus bus;
+    (* prefix="" *) interface SWDControllerPins pins;
+endinterface
+
+typedef struct {
+    Bit#(1) reset;
+    Bit#(3) reserved;
+    Bit#(1) apndp;
+    Bit#(1) rnw;
+    Bit#(2) addr;
+} Cmd deriving (Eq, Bits);
+
+typedef enum {
+    IDLE,
+    RESET,
+    RW
+} State deriving (Eq, Bits);
+
+(* synthesize *)
+module swd_module (ScaffoldSWDModule);
+    SWDController#(100) swd_controller <- mkSWDController();
+
+    Wire#(Bit#(16)) bus_address <- mkDWire(0);
+    Wire#(Bit#(8)) bus_write_data <- mkDWire(0);
+    Wire#(Bit#(1)) bus_write <- mkDWire(0);
+    Wire#(Bit#(1)) bus_read <- mkDWire(0);
+    Wire#(Bit#(1)) bus_en_rdata <- mkDWire(0);
+    Wire#(Bit#(1)) bus_en_wdata <- mkDWire(0);
+    Wire#(Bit#(1)) bus_en_cmd <- mkDWire(0);
+    Wire#(Bit#(1)) bus_en_status <- mkDWire(0);
+
+    Reg#(Bit#(8)) bus_reg_rdata <- mkReg(0);
+    Reg#(Bit#(8)) bus_reg_status <- mkReg(0);
+
+    Reg#(Bit#(2)) reg_rdata_cnt <- mkReg(0);
+
+    Reg#(Vector#(4, Bit#(8))) rdata <- mkReg(unpack(0));
+    Reg#(Status) status <- mkReg(unpack(0));
+    Reg#(Vector#(4, Bit#(8))) wdata <- mkReg(unpack(0));
+    Reg#(Maybe#(Cmd)) cmd <- mkReg(tagged Invalid);
+
+    Reg#(Bool) ready <- mkReg(False);
+    Reg#(State) state <- mkReg(IDLE);
+
+    rule do_bus_read (bus_read == 1);
+        case ({bus_en_rdata, bus_en_status}) matches
+            2'b10: 
+                begin 
+                    bus_reg_rdata <= rdata[reg_rdata_cnt];
+                    reg_rdata_cnt <= reg_rdata_cnt + 1;
```

Comment:
For reading `rdata`, you can simplify the circuit by using a shift register instead of indexing the vector (indexation is costly since it requires multiple muxes). There is no point of reading rdata twice, so this should be ok. Something like:

```
rule do_bus_read_rdata ((bus_read == 1) && (bus_en_rdata == 1) && (state != RW));
    bus_reg_rdata <= rdata[0];
        rdata <= shiftInAtN(rdata, 0);
    end
endrule
```

Note that there might be a conflict with RW state (taken into account in this code). I think it is not relevant to read `rdata` while a transaction is ongoing, so this should be fine.